stan/math/prim/mat/prob/neg_binomial_2_log_glm_lpmf.hpp

```diff
@@ -0,0 +1,195 @@
+
+      Array<T_partials_return, Dynamic, 1> n_arr(N, 1);
+      Array<T_partials_return, Dynamic, 1> phi_arr(N, 1);
+      {
+        scalar_seq_view<T_n> n_vec(n);
+        scalar_seq_view<T_precision> phi_vec(phi);
+        for (size_t n = 0; n < N; ++n) {
+          n_arr[n] = n_vec[n];
+          phi_arr[n] = value_of(phi_vec[n]);
+        }
+      }
+
+      Matrix<T_partials_return, Dynamic, 1> beta_dbl(M, 1);
+      {
+        scalar_seq_view<T_beta> beta_vec(beta);
+        for (size_t m = 0; m < M; ++m) {
+          beta_dbl[m] = value_of(beta_vec[m]);
+        }
+      }
+      Matrix<T_partials_return, Dynamic, Dynamic> x_dbl = value_of(x);
+
+      Array<T_partials_return, Dynamic, 1> theta_dbl = (x_dbl * beta_dbl
+      + Matrix<double, Dynamic, 1>::Ones(N, 1) * value_of(alpha)).array();
+      Array<T_partials_return, Dynamic, 1> log_phi = phi_arr.log();
+      Array<T_partials_return, Dynamic, 1> logsumexp_eta_logphi =
+        theta_dbl.binaryExpr(log_phi,
+                             [](T_partials_return xx, T_partials_return yy) {
+                               return log_sum_exp(xx, yy);
+                             });
+      Array<T_partials_return, Dynamic, 1> n_plus_phi = n_arr + phi_arr;
+
+      // Compute the log-density.
+      if (include_summand<propto>::value) {
+        logp -= (n_arr
+          + Array<double, Dynamic, 1>::Ones(N, 1))
+          .unaryExpr([](T_partials_return xx) {
+                       return lgamma(xx);
+                     }).sum();
+      }
+      if (include_summand<propto, T_precision>::value) {
+        logp += (phi_arr.binaryExpr(phi_arr, [](T_partials_return xx,
+                                                T_partials_return yy) {
+                                                  return multiply_log(xx, yy);
+                                              })
+                - phi_arr.unaryExpr([](T_partials_return xx) {
+                                      return lgamma(xx);
+                                    })).sum();
+      }
+      if (include_summand<propto, T_x, T_beta, T_alpha, T_precision>::value)
+        logp -= (n_plus_phi * logsumexp_eta_logphi).sum();
+      if (include_summand<propto, T_x, T_beta, T_alpha>::value)
+        logp += (n_arr * theta_dbl).sum();
+      if (include_summand<propto, T_precision>::value) {
+        logp += n_plus_phi.unaryExpr([](T_partials_return xx) {
+          return lgamma(xx);
+        }).sum();
+      }
+
+      // Compute the necessary derivatives.
+      operands_and_partials<T_x, T_beta, T_alpha,
+                            T_precision> ops_partials(x, beta, alpha, phi);
+
+      if (!(is_constant_struct<T_x>::value && is_constant_struct<T_beta>::value
+            && is_constant_struct<T_alpha>::value)) {
+        Matrix<T_partials_return, Dynamic, 1> theta_derivative(N, 1);
+        theta_derivative = (n_arr - (n_plus_phi / (phi_arr / (theta_dbl.exp())
+          + Array<double, Dynamic, 1>::Ones(N, 1)))).matrix();
```

Comment:
Do you have wisdom / rules of thumb to share on when `.matrix()` or `.eval()` are required?

Reply:
Sure, a little bit. .matrix() and .array() convert back and forth between Eigen::Matrix and Eigen::Array types. There is no runtime cost. The two types just provide a different interface: matrix operations vs coefficient-wise operations. Sometimes you need one, sometimes the other.
.eval() forces eager evaluation, as far as I understand it. By default, Eigen has lazy evaluation. I guess there are some cases where you want eager evaluation for efficiency or for correctness purposes, which you can force with .eval(). I haven't used it though.